src/integrationTest/java/com/mongodb/hibernate/junit/MongoExtension.java

```diff
@@ -0,0 +1,101 @@
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoDatabase;
+import com.mongodb.hibernate.internal.cfg.MongoConfigurationBuilder;
+import java.lang.reflect.Field;
+import java.util.Map;
+import java.util.function.Predicate;
+import org.bson.BsonDocument;
+import org.hibernate.cfg.Configuration;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.TestInstanceFactoryContext;
+import org.junit.jupiter.api.extension.TestInstancePreConstructCallback;
+import org.junit.jupiter.api.extension.TestInstancePreDestroyCallback;
+
+public final class MongoExtension
+        implements TestInstancePreConstructCallback,
+                TestInstancePreDestroyCallback,
+                BeforeAllCallback,
+                BeforeEachCallback {
+
+    private ExtensionContext outermostContext;
+
+    private MongoClient mongoClient;
+    private MongoDatabase mongoDatabase;
+
+    @Override
+    public void preConstructTestInstance(TestInstanceFactoryContext factoryContext, ExtensionContext context) {
+        if (factoryContext.getOuterInstance().isEmpty()) {
+            outermostContext = context;
+
+            @SuppressWarnings("unchecked")
+            Map<String, Object> hibernateProperties =
+                    (Map<String, Object>) (Map<?, Object>) new Configuration().getProperties();
+
+            var mongoConfig = new MongoConfigurationBuilder(hibernateProperties).build();
+            mongoClient = MongoClients.create(mongoConfig.mongoClientSettings());
+            mongoDatabase = mongoClient.getDatabase(mongoConfig.databaseName());
+        }
+    }
+
+    @Override
+    public void preDestroyTestInstance(ExtensionContext context) {
+        if (context.equals(outermostContext)) {
+            if (mongoClient != null) {
+                mongoClient.close();
+            }
+        }
+    }
+
+    @Override
+    public void beforeAll(ExtensionContext context) {
+        if (context.getTestInstance().orElse(null) instanceof MongoDatabaseAware mongoDatabaseAware) {
+            mongoDatabaseAware.injectMongoDatabase(mongoDatabase);
+        }
+
+        Predicate<Field> predicate = field -> MongoCollection.class.isAssignableFrom(field.getType());
+        findAnnotatedFields(context.getRequiredTestClass(), InjectMongoCollection.class, predicate)
+                .forEach(field -> {
+                    var annotation = field.getDeclaredAnnotation(InjectMongoCollection.class);
+                    String collectionName = annotation.name();
+                    var mongoCollection = mongoDatabase.getCollection(collectionName, BsonDocument.class);
+                    try {
+                        field.setAccessible(true);
+                        field.set(isStatic(field) ? null : context.getRequiredTestInstance(), mongoCollection);
```

Comment:
the above `context.getRequiredTestInstance()` won't throw exception for the precondition that some field is of instance field has implied that test instance has been available.